### PR TITLE
EL-4083 - Selection Issue

### PR DESCRIPTION
**Pull request by ashley-hunter**
#### Checklist

* [x] The contents of this PR are mine to submit. No third party code or other material is being committed.
* [x] New third party dependencies (if any) are linked via `package.json`. Third party dependencies are licensed under one of the following: Apache, MIT, BSD, Mozilla Public License, Eclipse Public License, or Oracle Binary Code License.
* [x] The code in this PR does not contain export-restricted encryption algorithms.

#### Ticket / Issue

https://portal.digitalsafe.net/browse/EL-4083

#### Description of Proposed Changes

The uxSelection directive and the uxSelectionItemDirective were both emitting far too often. The following changes have been made:

- Prevented the directives from emitting on init
- Prevented the directives from emitting when an input changes
- Prevented directives from emitting when an event has occurred but the value doesn't change
- Prevent directives from emitting until the output value has stabilized, for example, in row mode when a selection is made we first deselect all items, then select the newly selected item. The newly selected item should not emit the event when we initially deselect it.
- Added logic to ensure clicking on something like a checkbox emits as expected, e.g. when clicking a checkbox the selected input value may change, however the click on the row itself may also try and toggle the selected state, causing it to be toggled twice. I have added logic in there to prevent this double toggling.
- Added tests covering the scenarios

#### Documentation CI URL

https://pages.github.houston.softwaregrp.net/sepg-docs-qa/UXAspects_CI_UXAspects_EL-4083-Selection-Issue

#### Downstream Build(s)

* [caf/ux-aspects-micro-focus](https://jenkins.swinfra.net/job/SEPG/job/caf/job/caf~ux-aspects-micro-focus~EL-4083-Selection-Issue~CI/)

### CODE REVIEW

File: src/directives/selection/selection.directive.ts [hunk 1]
```diff
@@ -57,10 +58,10 @@ export class SelectionDirective<T> implements AfterContentInit, OnDestroy {
     private _onDestroy = new Subject<void>();
 
     /** Store the previous selection so we don't emit more than we have to */
-    private _lastSelection: ReadonlyArray<T>;
+    private _lastSelection: ReadonlyArray<T> = [];
 
     constructor(private _selectionService: SelectionService<T>, private _cdRef: ChangeDetectorRef) {
-        _selectionService.selection$.pipe(takeUntil(this._onDestroy), debounceTime(0)).subscribe(items => {
+        _selectionService.selection$.pipe(debounceTime(0), takeUntil(this._onDestroy)).subscribe(items => {
```
Comment (DeanHaire) [inline src/directives/selection/selection.directive.ts:64]: What's the purpose of `debounceTime(0)`, waiting for 0ms? Is it needed for some reason?

Comment (ashley-hunter) [reply src/directives/selection/selection.directive.ts:64]: Yes, so lets say we are in `row` mode, clicking on an item in row mode will select only that one row. What actually occurs is:
- deselect all rows
- select target row

Both of these actions will emit a new value on the observable. However the first is a transitional state, the other is the final state. There are many cases where this may fire multiple times before arriving on the final state.

The key point here that all these events fire synchronously the same tick of the event loop. 
`debounceTime(0)` will only proceed on a subsequent tick of the event loop, at which point all synchronous events that could have happened will have happened.

So, in short, the subscribe function will only get called after all the events have fired and the state is "stable".

Comment (DeanHaire) [reply src/directives/selection/selection.directive.ts:64]: That makes sense, thanks!

**Comment by DeanHaire** [comment]
Can you create a downstream build for this?

**Comment by ashley-hunter** [comment]
https://jenkins.swinfra.net/job/SEPG/job/caf/view/Developer/job/caf~ux-aspects-micro-focus~EL-4083-Selection-Issue~CI/1/

**Review by DeanHaire — APPROVED** [review]
Looks good, seems to work properly in each scenario I've tested 👍

**Review by donna-marijne — APPROVED** [review]
Thank you, looks good.
